dataclients/kubernetes/ingress.go

```diff
@@ -0,0 +1,775 @@
+	return algorithm
+}
+
+// TODO: find a nicer way to autogenerate route IDs
+func routeID(namespace, name, host, path, backend string) string {
+	namespace = nonWord.ReplaceAllString(namespace, "_")
+	name = nonWord.ReplaceAllString(name, "_")
+	host = nonWord.ReplaceAllString(host, "_")
+	path = nonWord.ReplaceAllString(path, "_")
+	backend = nonWord.ReplaceAllString(backend, "_")
+	return fmt.Sprintf("%s_%s__%s__%s__%s__%s", ingressRouteIDPrefix, namespace, name, host, path, backend)
+}
+
+// routeIDForCustom generates a route id for a custom route of an ingress
+// resource.
+func routeIDForCustom(namespace, name, id, host string, index int) string {
+	name = name + "_" + id + "_" + strconv.Itoa(index)
+	return routeID(namespace, name, host, "", "")
+}
+
+func setPath(m PathMode, r *eskip.Route, p string) {
+	if p == "" {
+		return
+	}
+
+	switch m {
+	case PathPrefix:
+		r.Predicates = append(r.Predicates, &eskip.Predicate{
+			Name: "PathSubtree",
+			Args: []interface{}{p},
+		})
+	case PathRegexp:
+		r.PathRegexps = []string{p}
+	default:
+		r.PathRegexps = []string{"^" + p}
+	}
+}
+
+func convertPathRule(
+	state *clusterState,
+	metadata *metadata,
+	host string,
+	prule *pathRule,
+	pathMode PathMode,
+) (*eskip.Route, error) {
+
+	ns := metadata.Namespace
+	name := metadata.Name
+
+	if prule.Backend == nil {
+		return nil, fmt.Errorf("invalid path rule, missing backend in: %s/%s/%s", ns, name, host)
+	}
+
+	var (
+		eps []string
+		err error
+		svc *service
+	)
+
+	var hostRegexp []string
+	if host != "" {
+		hostRegexp = []string{"^" + strings.Replace(host, ".", "[.]", -1) + "$"}
+	}
+	svcPort := prule.Backend.ServicePort
+	svcName := prule.Backend.ServiceName
+
+	svc, err = state.getService(ns, svcName)
+	if err != nil {
+		log.Errorf("convertPathRule: Failed to get service %s, %s, %s", ns, svcName, svcPort)
+		return nil, err
+	}
+
+	targetPort, err := svc.getTargetPort(svcPort)
+	if err != nil {
+		// fallback to service, but service definition is wrong or no pods
+		log.Errorf("Failed to find target port for service %s, fallback to service: %v", svcName, err)
+		err = nil
+	} else if svc.Spec.Type == "ExternalName" {
+		scheme := "https"
+		if targetPort != "443" {
+			scheme = "http"
+		}
+		u := fmt.Sprintf("%s://%s:%s", scheme, svc.Spec.ExternalName, targetPort)
+		f, e := eskip.ParseFilters(fmt.Sprintf(`setRequestHeader("Host", "%s")`, svc.Spec.ExternalName))
+		if e != nil {
+			return nil, e
+		}
+		return &eskip.Route{
+			Id:          routeID(ns, name, "", "", svc.Spec.ExternalName),
+			BackendType: eskip.NetworkBackend,
+			Backend:     u,
+			Filters:     f,
+		}, nil
+	} else {
+		// err handled below
+		eps, err = state.getEndpoints(ns, svcName, svcPort.String(), targetPort)
+		log.Debugf("convertPathRule: Found %d endpoints %s for %s", len(eps), targetPort, svcName)
+	}
+
+	// TODO: errors may get ignored if err != nil && err != errEndpointNotFound && len(eps) == 0
+	if len(eps) == 0 || err == errEndpointNotFound {
```

Comment:
similar to https://github.com/zalando/skipper/pull/1180/files#r369044301

Reply:
We found a bug, if servicePort is used with only referencing the name, the endpoint does not have a port number and you have to lookup from pods, which is a really bad choice in kubernetes.
The bug is not visible to the user, because we fallback to serviceIP if we can not get endpoints with port as number.
The question is, if we should change it anyways and document that we do not completely follow the K8s spec because of safety and efficiency reasons or we just let it like that.
wdyt @aryszka ?

Reply:
I think we should change it, but not in this PR, because of the additional testing needs. Currently, we fall back to using the service cluster IP and service port, which will cover those cases when the target port is referenced by value, however, will generate useless routes for those services that don't have endpoints, e.g. because of the deployment being scaled down to 0. In the future solution, we will need to distinguish between these two cases. Branch with a possible improvement can be found here: https://github.com/zalando/skipper/tree/improvement/service-port-fallback-handling